botocore/session.py

```diff
@@ -164,6 +165,7 @@ def _register_components(self):
         self._register_endpoint_resolver()
         self._register_event_emitter()
         self._register_response_parser_factory()
+        self._register_bucket_region_cache()
```

Comment:
Well for b we could add an event to client creation. a would also work, though I agree it could be dangerous. We could also just make it session based later.